app/interactors/add_coupon.rb

```diff
@@ -2,14 +2,23 @@ class AddCoupon
   include Interactor
 
   def call
-    return context.current_order.update(coupon: Coupon.find_by(code: context.code)) if coupon_valid?
+    return context.fail! if coupon_attached?
 
-    context.fail!
+    return context.fail! unless coupon_valid?
```

Comment:
two `returns` in the row is not a good idea. Maybe it would be better to make one method with validation?